### ISSUE TITLE
Environment.GetCommandLineArgs()[0] points to non-existing file in single-file apps

### ISSUE THREAD

**Issue by jkotas** [dotnet/runtime#40874]
Repro:
```csharp
using System;

class Program
{
    static void Main(string[] args)
    {
        Console.WriteLine(Environment.GetCommandLineArgs()[0]);
    }
}
```
1. `dotnet publish -r win-x64 --self-contained=false /p:PublishSingleFile=true`
2. `bin\Debug\net5.0\win-x64\publish\zzz.exe`

Actual result:
`c:\test\bin\Debug\net5.0\win-x64\publish\zzz.dll`

Expected result:
path to file that physically exists on disk

**Comment by am11** [comment]
Also, `typeof(Program).Assembly.Location` and `.CodeBase` return same absolute path to the non-existent `zzz.dll`.

**Comment by jkotas** [comment]
Related: #11305

**Comment by agocke** [comment]
@sbomer Could you take a look at this?

I'm actually not clear on what this should return. I think the user would expect the path to the single-file app, but if we don't return the path to the app host in the self-contained case it's also a little weird to do so now.

**Comment by mateoatr** [comment]
> but if we don't return the path to the app host in the self-contained case it's also a little weird to do so now.

But self-contained apps will deploy a dll that the user can actually see.

**Comment by agocke** [comment]
@mateoatr Right, that's why I'm leaning towards the file path of the host. It just seems like the only thing we could reasonably point to.

**Comment by agocke** [comment]
Actually @mateoatr could you look at this instead? I @sbomer is looking at https://github.com/dotnet/runtime/issues/36089

**Comment by swaroop-sridhar** [comment]
> I'm actually not clear on what this should return. I think the user would expect the path to the single-file app, but if we don't return the path to the app host in the self-contained case it's also a little weird to do so now.

When an app is executed using an apphost (self-contained, framework-dependent, or single-file), its intuitive to expect the host-path as `GetComandLineArgs[0]`. I'm not sure why the path to the DLL is returned in place of the host path. 

@jkotas Is this because we want to maintain consistency between runs that use `dotnet` host and apphost? 
Do we know of any tooling that depends on the fact that `GetComandLineArgs[0]` is the path to the main assembly? 

The other option is to return empty string as the assembly-path. The spec says "The first element in the array contains the file name of the executing program. If the file name is not available, the first element is equal to String.Empty."

> I'm leaning towards the file path of the host. It just seems like the only thing we could reasonably point to.

I second.

**Comment by jkotas** [comment]
> Is this because we want to maintain consistency between runs that use dotnet host and apphost?

I believe that the code to compute the 0-th command line arg was inherited from Windows Phone, and it happened to work ok for .NET Core up until now.

> Do we know of any tooling that depends on the fact that GetComandLineArgs[0] is the path to the main assembly?

I do not aware of any app that has dependency like this. It is likely that some exists. This falls into the general single-file compatibility issues. The apps that expect their managed entrypoint .dll exists on disk will need updates for single file.

> The spec says

The spec also says: "The program file name can, but is not required to, include path information."

> I'm leaning towards the file path of the host. 

+1
